plugins/grid-bookmark/src/GridBookmarkWidget/model.ts

```diff
@@ -51,106 +51,93 @@ export interface IExtendedLabeledRegionModel extends ILabeledRegionModel {
-            getSession(self).assemblyNames.includes(assembly),
-          ),
+          this.bookmarkAssemblies.filter(a => assemblyManager.get(a)),
         )
       },
     }))
     .views(self => ({
       get bookmarksWithValidAssemblies() {
-        return (
-          JSON.parse(
-            JSON.stringify(self.bookmarkedRegions),
-          ) as unknown as ILabeledRegionModel[]
-        ).filter(ele => self.validAssemblies.has(ele.assemblyName))
+        return self.bookmarkedRegions.filter(e =>
+          self.validAssemblies.has(e.assemblyName),
+        )
       },
     }))
     .views(self => ({
+      get sharedBookmarksModel() {
+        return SharedBookmarksModel.create({
+          sharedBookmarks: self.selectedBookmarks,
+        })
+      },
       get allBookmarksModel() {
+        // requires cloning bookmarks with JSON.stringify/parse to avoid duplicate reference to same object
+        // in the same state tree, will otherwise error when performing share
         return SharedBookmarksModel.create({
-          sharedBookmarks: self.bookmarksWithValidAssemblies,
+          sharedBookmarks: JSON.parse(
+            JSON.stringify(self.bookmarksWithValidAssemblies),
+          ),
         })
       },
     }))
-    .volatile(self => ({
-      selectedAssemblies: self.assemblies.filter((assembly: string) =>
-        getSession(self).assemblyNames.includes(assembly),
-      ),
-    }))
+
     .actions(self => ({
-      setSelectedAssemblies(assemblies: string[]) {
-        self.selectedAssemblies = assemblies
+      setSelectedAssemblies(assemblies?: string[]) {
+        self.selectedAssembliesPre = assemblies
       },
     }))
-    .actions(self => ({
-      updateSelectedAssembliesAfterClear() {
-        if (self.validAssemblies.size < self.selectedAssemblies.length) {
-          const rmAsm = self.selectedAssemblies.filter(
-            asm => !self.validAssemblies.has(asm),
-          )
-          rmAsm.forEach(asm => {
-            self.selectedAssemblies.splice(
-              self.selectedAssemblies.indexOf(asm),
-              1,
-            )
-          })
-          self.setSelectedAssemblies([...self.selectedAssemblies])
-        }
-      },
-      updateSelectedAssembliesAfterAdd() {
-        if (self.validAssemblies.size > self.selectedAssemblies.length) {
-          const newAsm = [...self.validAssemblies].filter(
-            asm => !self.selectedAssemblies.includes(asm),
-          )
-          self.setSelectedAssemblies([...self.selectedAssemblies, ...newAsm])
-        }
+
+    .views(self => ({
+      get selectedAssemblies() {
+        return (
+          self.selectedAssembliesPre?.filter(f =>
+            self.validAssemblies.has(f),
+          ) ?? [...self.validAssemblies]
+        )
       },
     }))
     .actions(self => ({
       importBookmarks(regions: Region[]) {
-        self.bookmarkedRegions = cast([...self.bookmarkedRegions, ...regions])
-        self.updateSelectedAssembliesAfterAdd()
```

Comment:
after this PR, selectedAssemblies tries to become a 'derived' state, so doesn't require manual update